federation/pkg/federation-controller/job/jobcontroller_test.go

```diff
@@ -0,0 +1,198 @@
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	batchv1 "k8s.io/kubernetes/pkg/apis/batch/v1"
+	kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
+	kubeclientfake "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5/fake"
+	"k8s.io/kubernetes/pkg/client/testing/core"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func installWatchReactor(fakeClien *core.Fake, resource string) {
+	fakeWatch := watch.NewRaceFreeFake()
+	fakeClien.PrependWatchReactor(resource, core.DefaultWatchReactor(fakeWatch, nil))
+	fakeClien.PrependReactor("create", resource, func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		obj := action.(core.CreateAction).GetObject()
+		fakeWatch.Add(obj)
+		return false, nil, nil
+	})
+	fakeClien.PrependReactor("update", resource, func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		fakeWatch.Modify(action.(core.UpdateAction).GetObject())
+		return false, nil, nil
+	})
+	fakeClien.PrependReactor("delete", resource, func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		fakeWatch.Delete(&batchv1.Job{
+			ObjectMeta: apiv1.ObjectMeta{
+				Name:      action.(core.DeleteAction).GetName(),
+				Namespace: action.GetNamespace(),
+			},
+		})
+		return false, nil, nil
+	})
+}
+
+func TestJobController(t *testing.T) {
+	flag.Set("logtostderr", "true")
+	flag.Set("v", "5")
+	flag.Parse()
+
+	jobReviewDelay = 50 * time.Millisecond
+	clusterAvailableDelay = 200 * time.Millisecond
+	clusterUnavailableDelay = 200 * time.Millisecond
+	allJobReviewDelay = 500 * time.Millisecond
+
+	fedclientset := fedclientfake.NewSimpleClientset()
+	installWatchReactor(&fedclientset.Fake, "jobs")
+
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-1", apiv1.ConditionTrue))
+	fedclientset.Federation().Clusters().Create(testutil.NewCluster("k8s-2", apiv1.ConditionTrue))
+
+	kube1clientset := kubeclientfake.NewSimpleClientset()
+	installWatchReactor(&kube1clientset.Fake, "jobs")
+	kube2clientset := kubeclientfake.NewSimpleClientset()
+	installWatchReactor(&kube2clientset.Fake, "jobs")
+
+	fedInformerClientFactory := func(cluster *fedv1.Cluster) (kubeclientset.Interface, error) {
+		switch cluster.Name {
+		case "k8s-1":
+			return kube1clientset, nil
+		case "k8s-2":
+			return kube2clientset, nil
+		default:
+			return nil, fmt.Errorf("Unknown cluster: %v", cluster.Name)
+		}
+	}
+	jobController := NewJobController(fedclientset)
+	fedjobinformer := testutil.ToFederatedInformerForTestOnly(jobController.fedJobInformer)
+	fedjobinformer.SetClientFactory(fedInformerClientFactory)
+
+	stopChan := make(chan struct{})
+	defer close(stopChan)
+	go jobController.Run(5, stopChan)
+
+	test := func(job *batchv1.Job) {
+		job, _ = fedclientset.Batch().Jobs(apiv1.NamespaceDefault).Create(job)
+		time.Sleep(1 * time.Second)
```

Comment:
@nikhiljindal Are you referring to deployment tests or to this code? In deployment tests there is limited number of loops - the channels are planted around so that you know when an update is made.